src/core/mesh/qgsmeshadvancedediting.cpp

```diff
@@ -591,3 +595,179 @@ bool QgsMeshEditRefineFaces::createNewBorderFaces( QgsMeshEditor *meshEditor,
+  bool calcZ = !mExpressionZ.isEmpty();
+  QgsExpression expressionX;
+  if ( calcX )
+  {
+    expressionX = QgsExpression( mExpressionX );
+    expressionX.prepare( &context );
+  }
+
+  QgsExpression expressionY;
+  if ( calcY )
+  {
+    expressionY = QgsExpression( mExpressionY );
+    expressionY.prepare( &context );
+  }
+
+  if ( calcX || calcY )
+  {
+    mNewXYValues.reserve( inputCount );
+    mOldXYValues.reserve( inputCount );
+  }
+
+  QgsExpression expressionZ;
+  if ( calcZ )
+  {
+    expressionZ = QgsExpression( mExpressionZ );
+    expressionZ.prepare( &context );
+    mNewZValues.reserve( inputCount );
+    mOldZValues.reserve( inputCount );
+  }
+
+  for ( int i = 0; i < mInputVertices.count(); ++i )
+  {
+    const int vertexIndex = mInputVertices.at( i );
+    context.lastScope()->setVariable( QStringLiteral( "_mesh_vertex_index" ), vertexIndex, false );
+
+    mChangingVertexMap[vertexIndex] = i;
+    const QVariant xvar = expressionX.evaluate( &context );
+    const QVariant yvar = expressionY.evaluate( &context );
+    const QVariant zvar = expressionZ.evaluate( &context );
+
+    const QgsMeshVertex &vert = mesh.vertex( vertexIndex );
+
+    if ( calcX || calcY )
+    {
+      mOldXYValues.append( QgsPointXY( vert ) );
+      mNewXYValues.append( QgsPointXY( vert ) );
+
+      const QList<int> facesAround = layer->meshEditor()->topologicalMesh().facesAroundVertex( vertexIndex );
+      concernedFaces.unite( qgis::listToSet( facesAround ) );
+    }
+
+    bool ok = false;
+    if ( calcX )
+    {
+      if ( xvar.isValid() )
+      {
+        double x = xvar.toDouble( &ok );
+        if ( ok )
+        {
+          mNewXYValues.last().setX( x );
+        }
+        else
+          return false;
+      }
+      else
+      {
+        return false;
+      }
+    }
+
+    if ( calcY )
+    {
+      if ( yvar.isValid() )
+      {
+        double y = yvar.toDouble( &ok );
+        if ( ok )
+        {
+          mNewXYValues.last().setY( y );
+        }
+        else
+          return false;
+      }
+      else
+        return false;
+    }
+
+    if ( calcZ )
+    {
+      if ( zvar.isValid() )
+      {
+        double z = zvar.toDouble( &ok );
+        if ( ok )
+        {
+          mNewZValues.append( z );
+          mOldZValues.append( vert.z() );
+        }
+        else
+          return false;
+      }
+      else
+        return false;
+    }
+  }
+
+  auto transformFunction = [this, layer ]( int vi )-> const QgsMeshVertex
+  {
+    return transformedVertices( layer, vi );
+  };
+
+  mNativeFacesIndexesGeometryChanged = qgis::setToList( concernedFaces );
+  return layer->meshEditor()->canBeTransformed( mNativeFacesIndexesGeometryChanged, transformFunction );
+}
+
+void QgsMeshTransformVerticesByExpression::setExpressions( const QString &expressionX, const QString &expressionY, const QString &expressionZ )
+{
+  mExpressionX = expressionX;
+  mExpressionY = expressionY;
+  mExpressionZ = expressionZ;
```

Comment:
don't you want to raise a signal that expression is changed?

Reply:
no, but maybe clear the data in the instance, the calculated vertices will be updated after calling `calculate()`